src/Symfony/Component/Validator/Constraints/WhenValidator.php

```diff
@@ -0,0 +1,60 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Validator\Constraints;
+
+use Symfony\Component\ExpressionLanguage\ExpressionLanguage;
+use Symfony\Component\Validator\Constraint;
+use Symfony\Component\Validator\ConstraintValidator;
+use Symfony\Component\Validator\Exception\LogicException;
+use Symfony\Component\Validator\Exception\UnexpectedTypeException;
+
+final class WhenValidator extends ConstraintValidator
+{
+    private ?ExpressionLanguage $expressionLanguage;
+
+    public function __construct(ExpressionLanguage $expressionLanguage = null)
+    {
+        $this->expressionLanguage = $expressionLanguage;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function validate(mixed $value, Constraint $constraint): void
+    {
+        if (!$constraint instanceof When) {
+            throw new UnexpectedTypeException($constraint, When::class);
+        }
+
+        $context = $this->context;
+        $variables = $constraint->values;
+        $variables['value'] = $value;
+        $variables['this'] = $context->getObject();
+
+        if ($this->getExpressionLanguage()->evaluate($constraint->expression, $variables)) {
+            $context->getValidator()->inContext($context)
+                ->validate($value, $constraint->constraints);
+        }
+    }
+
+    private function getExpressionLanguage(): ExpressionLanguage
+    {
+        if (null === $this->expressionLanguage) {
+            if (!class_exists(ExpressionLanguage::class)) {
+                throw new LogicException('Unable to use expressions as the Symfony ExpressionLanguage component is not installed.');
```

Comment:
To give some more context, what about mentioning the `When` validator somewhere in the message.

Reply:
we should also add the sentence that suggest running composer require (see similar messages in the codebase)